### PR TITLE
[feat]: AdaScale work with lr_scheduler and tests, examples

### PR DESCRIPTION
**Pull request by min-xu-ai**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/optim/adascale.py [hunk 1]
```diff
@@ -334,9 +340,16 @@ def zero_grad(self) -> None:
         return self._optimizer.zero_grad()
 
     def state_dict(self) -> Dict:
-        """Proxy function to optimizer, checkpointing needs this."""
+        """ Proxy function to optimizer, checkpointing needs this.
+
+            Note: Do NOT checkpoint in the middle of gradient accumulation since
```
Comment (blefaudeux) [inline fairscale/optim/adascale.py:345]: `.. note:` ? (see https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html)

File: tests/optim/test_single_node_adascale.py [hunk 2]
```diff
@@ -94,6 +94,9 @@ def test_grad_accum_gpu():
 def test_state_checkpointing():
     """ Test state checkpointing on GPU since that's the common case.
 
+        Note, we don't support checkpointing in the middle of gradient accumulation
```
Comment (blefaudeux) [inline tests/optim/test_single_node_adascale.py:97]: same

**Review by blefaudeux — APPROVED** [review]
LGTM, minor nitpicking on the rst syntax to get prettier html out of it, but that's all

File: fairscale/optim/adascale.py (hunk 1 above)
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:345]: Great idea! Will address both places!

File: README.md [hunk 3]
```diff
@@ -114,6 +114,21 @@ AdaScale can be used to wrap a SGD optimizer and to be used in DDP (Distributed
...
+from fairscale.optim import AdaScale
+
+...
+optim = AdaScale(SGD(model.parameters(), lr=0.1))
+...
+step = 0
+for sample in dataloader:
+    ...
+    step += optim.gain()
+    update_lr(step)  # This update_lr() function is unchanged from single GPU training.
```
Comment (mikerabbat) [inline README.md:128]: The standard PyTorch learning rate schedulers (in `torch.optim.lr_scheduler`) have a different syntax. Instead of passing `step` as an argument, they use a syntax like:
```
>>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
>>> scheduler = torch.optim.lr_scheduler.CyclicLR(optimizer, base_lr=0.01, max_lr=0.1)
>>> data_loader = torch.utils.data.DataLoader(...)
>>> for epoch in range(10):
>>>     for batch in data_loader:
>>>         train_batch(...)
>>>         scheduler.step()
```
I have an idea of how we could fix this pretty easily. We should be keeping track of the scale-invariant step number internally to AdaScale (accumulate the gain every time `step()`) is called. Then we could add a method to AdaScale that returns the number of times to call `scheduler.step()` after calling `optim.step()`.

One other point - the example should probably include the call to `optim.step()`

**Review by mikerabbat — COMMENTED** [review]
Not sure if this is a big deal, but the way learning rate scheduling is shown in the examples is different from how it is done with the standard PyTorch schedulers.

I don't know how standardized this is, though. Other packages (fairseq, classy_vision) appear to implement their own scheduling logic which is more similar to the one in the documentation.

File: README.md (hunk 3 above)
Comment (min-xu-ai) [reply README.md:128]: I see. That's a great observation! I didn't know about the standard way either. I will give this a try in the example. Stay tuned for new version of this PR later.

**Comment by min-xu-ai** [comment]
Hi, thanks for the suggestions reviewers! Can you take another look?

File: docs/requirements.txt [hunk 4]
```diff
@@ -3,4 +3,4 @@ sphinx==3.2.1
 sphinx_rtd_theme==0.4.3
 sphinxcontrib-programoutput==0.16
 git+https://github.com/pytorch/pytorch_sphinx_theme.git#egg=pytorch_sphinx_theme
-torch==1.6.0
+torch>=1.6.0
```
Comment (blefaudeux) [inline docs/requirements.txt:6]: oh, nice catch

File: fairscale/optim/adascale.py [hunk 5]
```diff
@@ -38,27 +38,36 @@
...
         adascale = AdaScale(optim)
 
+        step = 0
         for epoch in ...:
             for batch in ...:
                 optim.zero_grad()
                 loss = ...
                 loss.backward()
+                step += adascale.gain()
+                update_lr(step)
```
Comment (mikerabbat) [inline fairscale/optim/adascale.py:70]: Would it make sense to use the same example code across Readme, rst, and pydoc?

File: fairscale/optim/adascale.py [hunk 6]
```diff
@@ -38,27 +38,36 @@
 import torch
 from torch.autograd import Variable
 import torch.distributed as dist
+from torch.optim import Optimizer
 
 
-class AdaScale(object):
+class AdaScale(Optimizer):
```
Comment (mikerabbat) [inline fairscale/optim/adascale.py:44]: Not sure this is needed.

In the documentation example in the rst file, things are instantiated in the following order:
```
optimizer = [base optimizer class]
scheduler = ...
optimizer = AdaScale(optimizer)
```

As long as things are done in that order, shouldn't everything work ok without needing AdaScale to subclass Optimizer?

Comment (mikerabbat) [reply fairscale/optim/adascale.py:44]: Looks like things happen in a different order in `test_lr_scheduler`

**Review by mikerabbat — COMMENTED** [review]
Few additional comments/suggestions

File: README.md [hunk 7]
```diff
@@ -117,6 +117,31 @@ AdaScale can be used to wrap a SGD optimizer and to be used in DDP (Distributed
...
+optim = AdaScale(SGD(model.parameters(), lr=0.1))
+scheduler = LambdaLR(optim, ...)
+...
+# Note: the train loop should be with DDP or with gradient accumulation.
+last_epoch = 0
+step = 0
+done = False
+while not done:
+    for sample in dataset:
+        ...
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply README.md:135]: Good catch! Will do.

File: docs/source/tutorials/adascale.rst [hunk 8]
```diff
@@ -75,18 +79,30 @@ as follows.
...
                 outputs = model(data)
                 loss = loss_fn(outputs, target)
                 loss.backward()
+                step += optimizer.gain()
                 optimizer.step()
+                epoch = step // len(dataloader)
+                if last_epoch != epoch:
+                    scheduler.step()
+                    last_epoch = epoch
+                if epoch >= len(epochs):
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply docs/source/tutorials/adascale.rst:107]: Right! Sharp eyes! Will fix.

File: fairscale/optim/adascale.py (hunk 5 above)
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:70]: I was thinking that with a varieties of examples, maybe it helps the user more in different situations?

File: fairscale/optim/adascale.py (hunk 6 above)
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:44]: I see. I meant the order in the test but I both orders are possible. Which one do you think is cleaner? It seems that if we only support a specific order, it is harder for people to understand the error when they reversed the order? What's the potential downside of subclassing from Optimizer?

File: fairscale/optim/adascale.py (hunk 5 above)
Comment (mikerabbat) [reply fairscale/optim/adascale.py:70]: Sure, makes sense, though it could also potentially cause confusion. I'd suggest if we want to show multiple examples, to group them together in the same place (example 1, example 2, ...) and clearly point out that there's more than one way to use this.

Comment (min-xu-ai) [reply fairscale/optim/adascale.py:70]: Good idea. How about the following?

* we keep the README.md example more "official" (the lr_schedule one)
* keep the tutorial example as is
* here, we have two examples, one for each style

File: fairscale/optim/adascale.py (hunk 6 above)
Comment (mikerabbat) [reply fairscale/optim/adascale.py:44]: The main downside I could imagine is that we don't currently wrap or implement all methods in the Optimizer interface. If someone uses one of the methods we don't explicitly wrap or override, it wouldn't have the expected effect, since it will operate on the adascale instance rather than adascale._optimizer. I don't have a strong opinion, though. Is it strange to have a class subclass Optimizer, but then internally have an instance of Optimizer that is really providing the main implementation?

File: fairscale/optim/adascale.py (hunk 5 above)
Comment (mikerabbat) [reply fairscale/optim/adascale.py:70]: Sounds great!

File: docs/requirements.txt (hunk 4 above)
Comment (msbaines) [reply docs/requirements.txt:6]: This file is used by the readthedocs service. The impact of this is that the readthedocs service will now use latest PyTorch after every commit to generate our documentation. That might be OK since we now try to support latest PyTorch ASAP.

**Review by msbaines — APPROVED** [review]
LGTM. Thanks for writing docs and tests.

File: fairscale/optim/adascale.py (hunk 6 above)
Comment (min-xu-ai) [reply fairscale/optim/adascale.py:44]: We discussed more over chat. We will keep the subclass for now so that we maintain to be a full wrapper.

**Review by mikerabbat — APPROVED** [review]
Looks good to me. Thanks for taking all the feedback and suggestions into account!
